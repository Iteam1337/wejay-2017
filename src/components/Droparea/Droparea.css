/* ==================================================================
   Droparea
   ================================================================== */

.Droparea {
  align-items: center;
  background: red;
  border: 0;
  box-sizing: border-box;
  color: transparent;
  display: flex;
  font-family: 'Roboto';
  font-size: 32px;
  font-weight: 300;
  flex-direction: column;
  height: 200px;
  justify-content: center;
  letter-spacing: 0.02em;
  padding: 80px 0;
  position: fixed;
  resize: none;
  left: 0;
  text-align: center;
  bottom: 0;
  transition: all 200ms ease-in-out;
  z-index: 20;
  user-select: none;
  width: 100%;

  &:focus {
    outline: none;
  }

  @media (max-height: 1100px) {
    height: 120px;
    padding: 40px 0;
  }

  @media (--bp-med) {
    box-sizing: border-box;
  }
}

.active {
  background: var(--color-cod-gray);
  border-color: var(--color-cod-gray) !important;
  text-shadow: 0 0 0 #f0f0f0;
}

.Droparea__wrap {
  position: relative;
}

.icon {
  color: #fff;
  font-size: 70px;
  position: absolute;
  left: 50%;
  top: 32%;
  transform: translateX(-50%);
  z-index: 100;
}

.empty {
  border: 3px dashed #444;
  border-radius: 4px;
  font-size: 21px;
  line-height: 1.5;
  padding: 250px 30px 0;
  position: relative;
  height: 500px;
  text-shadow: 0 0 0 #f0f0f0;
  width: 800px;
}
