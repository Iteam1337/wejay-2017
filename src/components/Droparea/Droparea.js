import './Droparea.css'
import React, { Component } from 'react'
import { gql, graphql } from 'react-apollo'
import { roomQuery } from 'views/Room/Room'
import classnames from 'classnames'
import md5 from 'md5'

type DropareaProps = {
  message?: string,
  mutate: Function,
  roomName: string
}

class Droparea extends Component {
  props: DropareaProps

  state = {
    isDragOver: false
  }

  _onDragEnter = e => {
    if (e) {
      e.preventDefault()
    }

    this.setState({
      isDragOver: true
    })
  }

  _onDragLeave = e => {
    if (e) {
      e.preventDefault()
    }

    this.setState({
      isDragOver: false
    })
  }

  _onDrop = e => {
    const { mutate, roomName } = this.props
    const userId = md5(localStorage.getItem('user'))

    e.stopPropagation()
    e.preventDefault()

    const songs = e.dataTransfer
      .getData('text')
      .replace(/https:\/\/open.spotify.com\/track\//gi, '')
      .split('\n')

    songs.reverse().forEach(track => {
      mutate({
        variables: {
          input: {
            roomName,
            userId,
            spotifyId: track
          }
        },
        optimisticResponse: {
          queueTrack: {
            album: [],
            artists: [
              {
                name: 'Loading',
                __typename: 'Artist'
              }
            ],
            duration: 0,
            name: 'Awesome track',
            spotifyUri: track,
            started: 0,
            user: {
              email: '',
              id: userId,
              __typename: 'User'
            },
            __typename: 'Track'
          }
        },
        update: (store, { data: { queueTrack } }) => {
          const data = store.readQuery({
            query: roomQuery,
            variables: {
              name: roomName
            }
          })

          const exists = data.room.queue.find(
            track => track.spotifyUri === queueTrack.spotifyUri
          )

          const notCurrent =
            data.room.currentTrack &&
            data.room.currentTrack.spotifyUri !== queueTrack.spotifyUri

          if (!exists && notCurrent) {
            data.room.queue.push(queueTrack)
          }

          store.writeQuery({
            query: roomQuery,
            variables: {
              name: roomName
            },
            data
          })
        }
      })
    })

    this.setState({
      isDragOver: false
    })
  }

  render () {
    return (
      <div className="Droparea__wrap">
        <textarea
          className={classnames('Droparea', {
            'Droparea--is-dragging': this.state.isDragOver
          })}
          defaultValue={this.props.message}
          onDragEnter={this._onDragEnter}
          onDragLeave={this._onDragLeave}
          onDrop={this._onDrop}
          spellCheck={false}
        />
      </div>
    )
  }
}

Droparea.defaultProps = {
  message: 'Drop it like it\'s hot!'
}

const addTrackMutation = gql`
  mutation queueTrack($input: QueueInput!) {
    queueTrack(input: $input) {
      album {
        images {
          url
        }
        name
      }
      artists {
        name
      }
      duration
      name
      spotifyUri
      user {
        email
        id
      }
    }
  }
`

export default graphql(addTrackMutation)(Droparea)
