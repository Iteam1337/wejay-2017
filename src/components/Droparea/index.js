import Droparea from './Droparea'
export default Droparea
