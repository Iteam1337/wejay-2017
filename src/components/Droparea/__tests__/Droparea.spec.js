import React from 'react'
import { shallow } from 'enzyme'
import { Droparea } from '../Droparea'

describe('components/Droparea', () => {
  let component
  let event
  let instance

  beforeEach(() => {
    event = {
      preventDefault: jest.fn()
    }

    component = shallow(<Droparea />)

    instance = component.instance()
  })

  it('renders Droparea', () => {
    expect(component).toMatchSnapshot()
  })

  describe('#onDragEnter', () => {
    it('should prevent default if an event is provided', () => {
      instance.onDragEnter(event)

      expect(event.preventDefault).toHaveBeenCalled()
    })

    it('should set dragover on state', () => {
      instance.onDragEnter()

      expect(event.preventDefault).not.toHaveBeenCalled()
      expect(component.state().isDragOver).toBe(true)
    })
  })

  describe('#onDragLeave', () => {
    it('should prevent default if an event is provided', () => {
      instance.onDragLeave(event)

      expect(event.preventDefault).toHaveBeenCalled()
    })

    it('should reset dragover on state', () => {
      instance.onDragLeave()

      expect(event.preventDefault).not.toHaveBeenCalled()
      expect(component.state().isDragOver).toBe(false)
    })
  })
})
