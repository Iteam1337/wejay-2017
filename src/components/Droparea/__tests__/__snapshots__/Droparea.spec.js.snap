// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Droparea renders Droparea 1`] = `
<div
  className="Droparea__wrap"
>
  <textarea
    className="Droparea"
    defaultValue="Drop it like it's hot!"
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDrop={[Function]}
    spellCheck={false}
  />
</div>
`;
